### ISSUE TITLE
[kafka-binder] Problem handling errors for a kafka consumer in batch mode

### ISSUE THREAD

**Issue by ferblaca** [spring-cloud/spring-cloud-stream#2632]
**Describe the issue**
It seems that by throwing a `BatchListenerFailedException` in the consumer logic in Batch mode, the exception is not interpreted as such and therefore does not do what is expected.
In the [demo application you can access here](https://github.com/ferblaca/demoKafkaBatch), a Kafka consumer is configured in Batch mode with default values. 
The Consumer receives a list of 10 String messages, and on reading the sixth message it throws a `BatchListenerFailedException` exception. 

```
    @Bean
    public Consumer<List<String>> consumerBatch() {
        return s -> {
            s.stream().forEach(s1 -> {
                LOG.info(s1);
                if (s1.equals("test-5")) {
                    throw new BatchListenerFailedException(s1, 5);
                }
            });
        };
    }
```

This is the trace of the exception:
```
2023-01-24T16:29:22.977+01:00 ERROR 19856 --- [container-0-C-1] o.s.k.l.FallbackBatchErrorHandler        : Records discarded: destination.topic-0@0,destination.topic-0@1,destination.topic-0@2,destination.topic-0@3,destination.topic-0@4,destination.topic-0@5,destination.topic-0@6,destination.topic-0@7,destination.topic-0@8,destination.topic-0@9
```

**To Reproduce**
1. Run the [demo application](https://github.com/ferblaca/demoKafkaBatch).
2. Debug [this line](https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java#L148) of `FailedBatchProcessor` class.
3. It is noted that the `getBatchListenerFailedException()` method does not retrieve the `BatchListenerFailedException`.

![image](https://user-images.githubusercontent.com/9643385/214335934-c88c5f3e-3532-47ae-a85d-921478ec07cb.png)

> The problem is that once all retries are exhausted (9 by default) all records are committed in a single poll()

**Version of the framework**
Spring Cloud 2022.0.0
Spring Boot 3.0.2

**Expected behavior**
A partial commit is performed on the first 5 Consumer Records and a seek is performed on the other 5 after the retries have been exhausted.

**Additional context**
At least with Spring Cloud version 2021.0.2 it works as expected.

**Comment by garyrussell** [comment]
Not sure if it is an SCSt problem or Spring Integration problem, but we seem to have lost the root cause...

2021.0.2...

<img width="985" alt="Screenshot 2023-01-24 at 5 52 09 PM" src="https://user-images.githubusercontent.com/483832/214439407-551eabad-1254-47f2-ab80-ec6e8bd74613.png">

`LEFE -> MessageHandlingException -> BLEFE`

2022.0.0

<img width="955" alt="Screenshot 2023-01-24 at 5 54 23 PM" src="https://user-images.githubusercontent.com/483832/214439428-091b6f3f-3423-49d9-a8dc-6709b4a22788.png">

`LEFE -> MessageDeliveryException -> MessageDispatchingException -> null` (null shows as a self reference in the stack trace).

The stack trace is lost because a `Dispatcher Has No Subscribers` `MessageDispatchingException` has no cause.

cc/ @olegz @artembilan

**Comment by artembilan** [comment]
I think it is not correct to expect to see an `ErrorMessage` content connected to the current stack trace. 
It is just a message for a channel we send and if we fail there that is not a responsibility of a payload of this `ErrorMessage` to be added somehow to the cause of `Dispatcher Has No Subscribers`.
We do have that though when we go to the `failedMessage` and its payload since it is an `ErrorMessage` by itself. But that's already out of Java stacktrace scope.

**Comment by garyrussell** [comment]
I agree - especially if the DHNS occurs in user code.

The root cause is that the channel is the wrong type; when there is no subscriber, it should be a no-op, which SCSt responds with re-throwing the original exception to the container (which is required for proper functioning in this scenario).

**Comment by garyrussell** [comment]
>I agree - especially if the DHNS occurs in user code.

Although we could add the payload to the `suppressedExceptions` - but that wouldn't help in this case, because the error handler doesn't look at them.

**Comment by artembilan** [comment]
And I guess you mean it was like that when error channel was of a `PublishSubscribeChannel` type before.
Or the `minSubscribers` has been set not to `0` somehow...

**Comment by garyrussell** [comment]
Right. Something must have changed; it also has that qualifier now `2120532393`.

**Comment by garyrussell** [comment]
Running in a debugger and `integrationProperties.isErrorChannelRequireSubscribers()` is `true` :(

**Comment by garyrussell** [comment]
In previous versions, the binder error channel was bridged to the global error channel; hence `BinderErrorChannel` coerces this to `true`. 

It looks like the bridge is no longer a subscriber; hence the error.

**Comment by artembilan** [comment]
Yeah... That's correct. That's a default one. See docs: https://docs.spring.io/spring-integration/reference/html/configuration.html#global-properties
However that applies for a global `errorChannel` bean. But here it looks like a binding-specific channel for error handling.
Either way the global `errorChannel` bean has one `LoggingHandler` subscriber by default 🤷

**Comment by garyrussell** [comment]
Exactly, so something is broken in the binder.

**Comment by garyrussell** [comment]
Arrrrggggghhhhh...

Known problem; fixed in 4.0.1 (2022.0.1).

https://github.com/spring-cloud/spring-cloud-stream/issues/2610

https://github.com/spring-cloud/spring-cloud-stream/commit/686a75c077c6ac109e85cea06c5e08e9cde89d04

**Comment by garyrussell** [comment]
But I had to patch the test code

```java
    @Bean
    public Consumer<List<String>> consumerBatch() {
        return s -> {
        	AtomicInteger i = new AtomicInteger();
            s.stream().forEach(s1 -> {
                LOG.info(s1);
                if (s1.equals("test-5")) {
                    throw new BatchListenerFailedException(s1, i.get());
                }
                i.incrementAndGet();
            });
        };
    }
```

Previously, the index was hard coded to 5 instead of using the exact index of `test-5`, which caused this https://github.com/spring-projects/spring-kafka/issues/2560
